File: measurements/workloads/sort/qsort.java
import java.util.Scanner;
import java.lang.management.ThreadMXBean;
import java.lang.management.ManagementFactory;

class Sort {
	private ThreadMXBean timer = ManagementFactory.getThreadMXBean();
	public void quicksort(long[] data, int offset, int count) {
		if (count < 2) {
			return;
		}

		long pivot = data[offset + count / 2];

		int i, j;
		for (i = 0, j = count - 1; ; i++, j--) {
			while (data[offset + i] < pivot) i++;
			while (data[offset + j] > pivot) j--;

			if (i >= j) {
				break;
			}

			long tmp = data[offset + i];
			data[offset + i] = data[offset + j];
			data[offset + j] = tmp;
		}

		quicksort(data, offset, i);
		quicksort(data, offset + i, count - i);
	}

	public void run() {
		long beforeCpu = timer.getCurrentThreadCpuTime();
		long beforeWall = System.nanoTime();
		long startWall = System.currentTimeMillis();

		Scanner scanner = new Scanner(System.in);
		int length = Integer.parseInt(scanner.nextLine());

		long data[] = new long[length];
		int i = 0;

		while (scanner.hasNext()) {
			data[i] = Long.parseLong(scanner.nextLine());
			i += 1;
		}

		quicksort(data, 0, length);

		for (i = 0; i < length; i++) {
			System.out.println(data[i]);
		}

		long afterCpu = timer.getCurrentThreadCpuTime();
		long afterWall = System.nanoTime();

		System.err.printf("start-wall,%.9f%n", startWall / 1e3);
		System.err.printf("cpu,%.9f%n", ((double) (afterCpu - beforeCpu)) / 1e9);
		System.err.printf("wall,%.9f%n", ((double) (afterWall - beforeWall)) / 1e9);
	}

	public static void main(String[] args) {
		Sort sort = new Sort();
		sort.run();
	}
}

File: measurements/workloads/sort/qsort.py
#!/usr/bin/env python3

import sys
import time

def quicksort(data, offset, count):
	if count < 2:
		return

	pivot = data[offset + count // 2]

	i = 0
	j = count - 1

	while True:
		while data[offset + i] < pivot:
			i += 1

		while data[offset + j] > pivot:
			j -= 1

		if i >= j:
			break

		tmp = data[offset + i]
		data[offset + i] = data[offset + j]
		data[offset + j] = tmp

		i += 1
		j -= 1
	
	quicksort(data, offset, i)
	quicksort(data, offset + i, count - i)

before_cpu = time.process_time()
before_wall = time.time()

lines = iter(sys.stdin)
length = int(next(lines))
data = []

for line in lines:
	data.append(int(line))

quicksort(data, 0, length)

for item in data:
	print(item)

after_cpu = time.process_time()
after_wall = time.time()

print(f"start-wall,{before_wall:.9f}", file=sys.stderr)
print(f"cpu,{(after_cpu - before_cpu):.9f}", file=sys.stderr)
print(f"wall,{(after_wall - before_wall):.9f}", file=sys.stderr)

